rules/scripts/bitstreams_workspace.py

```diff
@@ -208,40 +211,50 @@ def _GetFromRemote(self, key):
         if latest:
             self.Touch(latest)
 
-    def GetFromCache(self, key):
+    def GetFromCache(self, key: str) -> Dict[str, Set[str]]:
         """Get the requested bitstream files.
 
         Args:
-            key: str; A git hash or the string 'latest'.
+            key: A git hash or the string 'latest'.
         Returns:
-            dict[str:str]: A dictionary mapping the bitstream types to
-                           their target files.
+            A dictionary that maps file extensions to the set of bitstream files
+            with that extension.
         """
         files = self._GetFromLocal(key)
         if not files:
             self._GetFromRemote(key)
             files = self._GetFromLocal(key)
-        return {os.path.splitext(f)[1][1:]: f for f in files}
 
-    def WriteBuildFile(self, build, key):
-        """Write a BUILD file for the requested bitstream files.
+        extension_map: Dict[str, Set[str]] = collections.defaultdict(set)
+        for file_name in files:
+            _, extension = os.path.splitext(file_name)
+            if extension[0] != '.':
+                logging.error("Bitstream file has no extension: " +
+                              repr(file_name))
+                sys.exit(1)
+            key = extension[1:]
+            extension_map[key] |= {file_name}
+        return extension_map
 
-        Args:
-            build: path; Filename of the BUILD file to write.
-            key: str; A git hash or the string 'latest'.
-        """
-        param = self.GetFromCache(key)
-        param['datetime'] = datetime.datetime.now().isoformat()
-        param['key'] = key
+    @staticmethod
+    def _GetDateTimeStr():
+        return datetime.datetime.now().isoformat()
 
-        if 'orig' not in param or 'splice' not in param:
+    def _ConstructBazelString(self, build_file: Path, key: str) -> str:
+        files_by_extension = self.GetFromCache(key)
+
+        if len(files_by_extension.get('orig', set())) != 1 or \
+           len(files_by_extension.get('splice', set())) != 1:
             logging.error(
                 """Could not find the bitstreams to generate a BUILD file:{}
-in params:{}
-using key:{}""".format(build, param, key))
+in files_by_extension:{}
+using key:{}""".format(build_file, files_by_extension, key))
             sys.exit(1)
-        with open(build, 'wt') as f:
-            f.write("""# This file was autogenerated. Do not edit!
+
+        (test_rom_file, ) = files_by_extension['orig']
+        (mask_rom_file, ) = files_by_extension['splice']
+
+        bazel_string = """# This file was autogenerated. Do not edit!
```

Comment:
I mildly dislike giant triple-quote strings in Python because they break the visual flow when you're following indentation.

My usual pattern is to append each line to a list and then do `'\n'.join(lines)` at the end. Any objections to doing that in a cleanup PR?

Reply:
This is a historic error on my part.  Usually when I use triple-quoted strings, I make them constants at the top of the file.
```
_BUILD_FILE = """# This file was auto generated.
<stuff>
"""
```

I usually use the `join` approach when things are less-constant and the triple-quoted string at the top of the file when things can be easily templated with `CONST_STRING.format(**some_dict)`.

Reply:
OK great, we're on the same page. I declare this out of scope for this CL, but I'll try to factor out the large triple quote literals from bitstreams_workspace.py in another PR.